### DISCUSSION TITLE
AptX LL on A2DP Source

### DISCUSSION THREAD

**Discussion by jonrichings** [pschatzmann/ESP32-A2DP#216] [Q&A] [answered]
Hi Phil, great admirer of your work :)  

I am just embarking on an ESP32-based synth design (maybe to become commercial).  I have some forward looking questions that will guide my approach:

Do you think adding AptX LL to the A2DP source would be even possible?  Could you be encouraged to do it ;)
also
Are you aware of the Faust (Functional Audio Stream) programming language?  Do you think it could be integrated with your Arduino Audio Tools?  I'm leaning towards Faust for the synthesis, mostly because of their Karplus-Strong and Physical Modelling libs.

TIA, Jon

**Comment by pschatzmann** [comment]
AptX: Since I prefer to support different environments, I don't think this will be possible and you would really need to convince Espressif to support this.  

I have Faust on my to do list and it should be quite easy to add some basic support, if you are ready to perform some slight modifications to the generate source code. Like renaming the generated cpp to h and including an additional header file. 
To generate the code you would select platform Source and Architecture cpp. Please note that anything which uses a lot of memory will not work.

Would you volunteer to test ?

**Reply by jonrichings** [reply]
Absolutely, yes.  I'm only just setting up a LyraT board to test code with so maybe a day or two away from testing (also have a hardware config issue I ask about in another thread).  I'll be very happy to test as soon as I can.

As for AptX.  I thought that might be the answer.  My fallback is to use an external BT module that has AptX.  From what I've heard from Espressif's forums, AptX is not likely.

**Comment by pschatzmann** [comment]
Here is a first [example as source](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/streams-faust-i2s) and another as [signal processor](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/streams-generator-faust-i2s).

See also my [blog](https://www.pschatzmann.ch/home/2022/04/22/using-faust-dsp-with-my-arduino-audio-tools/)

ps. Please note that the examples above have been [moved in the meantime](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-faust)

**Comment by jokubasver** [comment]
There's this, but it's for sink: https://github.com/cfint/esp32-a2dp-sink

**Reply by jonrichings** [reply]
Thanks.  I saw that too which is why I asked Phil.  But it's sink only and I need source.

**Comment by pschatzmann** [comment]
As far as I understand this is using a custom patched esp-idf version. So this is nothing which would work in Arduino or PlatformIO.
Please correct me if I am wrong...

**Comment by pschatzmann** [comment, marked as answer]
IDF 6.1 is supporting different codecs (see https://github.com/pschatzmann/ESP32-A2DP/discussions/805)
